ocaml/xapi/nm.ml

```diff
@@ -333,10 +334,25 @@ let bring_pif_up ~__context ?(management_interface=false) (pif: API.ref_PIF) =
         begin match rc.API.pIF_bond_master_of with
           | [] -> ()
           | bond :: _ ->
-            let slaves = Db.Bond.get_slaves ~__context ~self:bond in
+            let dbBond = Db.Bond.get_record ~__context ~self:bond in
+
+            if dbBond.API.bond_auto_update_mac then begin
+              (* Make sure the MAC address of the PIF Bond master matches the MAC address of the Bond's primary slave *)
+              let primarySlaveRef = dbBond.API.bond_primary_slave in
+              let primarySlaveMAC = Db.PIF.get_MAC ~__context ~self:primarySlaveRef in
+              if rc.pIF_MAC <> primarySlaveMAC then begin
+                Db.PIF.set_MAC ~__context ~self:pif ~value:primarySlaveMAC;
+                mac_changed := true;
+              end;
+            end;
+
+            let slaves = dbBond.API.bond_slaves in
+
             List.iter (fun self -> Db.PIF.set_currently_attached ~__context ~self ~value:false) slaves
         end;
 
+        let rc = if !mac_changed then Db.PIF.get_record ~__context ~self:pif else rc in
```

Comment:
An alternative to eliminate the `get_record` call and the `ref` is to "return" an updated PIF record from the above `match` statement. A structure like:
```
let rc =
  match rc.API.pIF_bond_master_of with
  | [] -> rc
  | bond :: _ ->
    ...
    List.iter ...slaves;
    if ... then
      {rc with mAC = primarySlaveMAC}
    else
      rc
in
```